### PR TITLE
allowing `Matches` to accept `re.compile`

### PR DESCRIPTION
**Pull request by bandophahita**
The resolving function `matches_regexp` of `Matches` can accept either `str` or `Pattern`.  This update makes it possible to pass the pattern directly into `Matches`.

### CODE REVIEW

File: screenpy/resolutions/matches.py [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
+
+from re import Pattern
+from typing import TYPE_CHECKING
+
 from hamcrest import matches_regexp
-from hamcrest.core.matcher import Matcher
 
 from screenpy.pacing import beat
 
+if TYPE_CHECKING:  # pragma: no cover
```
Comment (perrygoy) [inline screenpy/resolutions/matches.py:12]: What is pragma? Do we use it?

File: screenpy/resolutions/matches.py [hunk 2]
```diff
@@ -17,14 +24,21 @@ class Matches:
         )
     """
 
+    @property
+    def item_to_log(self) -> str:
+        """Represent the item in a log-friendly way."""
+        if isinstance(self.pattern, Pattern):
+            return f"r'{self.pattern.pattern}'"
+        return f"r'{self.pattern}'"
```
Comment (perrygoy) [inline screenpy/resolutions/matches.py:37]: Do we want to move this logic to `represent_prop`?

File: screenpy/resolutions/matches.py [hunk 3]
```diff
@@ -17,14 +24,21 @@ class Matches:
         )
     """
 
+    @property
+    def item_to_log(self) -> str:
```
Comment (perrygoy) [inline screenpy/resolutions/matches.py]: I think i'd prefer this to be called `pattern_to_log` :P

File: screenpy/resolutions/matches.py [hunk 4]
```diff
@@ -17,14 +24,21 @@ class Matches:
         )
```
Comment (perrygoy) [inline screenpy/resolutions/matches.py:24]: Can you add more examples with the new options, like `re.compile` and a plain string?

**Review by perrygoy — CHANGES_REQUESTED** [review]
Some changes requested and a couple questions!

File: screenpy/resolutions/matches.py (hunk 1 above)
Comment (bandophahita) [reply screenpy/resolutions/matches.py:12]: It's the comment to tell `coverage` "this section is not to be counted".  There is still an [open issue](https://github.com/nedbat/coveragepy/issues/831) to automatically ignore `if TYPE_CHECKING`.  I have been slowly adding tests based on `coverage` reports.

I've got a code coverage branch that I'll be submitting after this PR with a ton of lines just like this.

File: screenpy/resolutions/matches.py (hunk 3 above)
Comment (bandophahita) [reply screenpy/resolutions/matches.py]: Funny, I almost used that.  I think I was trying to find a universal name for the property that would work for all of the classes regardless of what the "item" is.  Not sure why I thought that was important.
Updating it.

File: screenpy/resolutions/matches.py (hunk 2 above)
Comment (bandophahita) [reply screenpy/resolutions/matches.py:37]: There is a slight issue with doing that.  I thought it best to show the pattern in the log as `r'pattern'`, which includes the `r` before the string quotes. 

`... hoping it's text matching the pattern r'\tpattern'.` -- GOOD
`... hoping it's text matching the pattern '\\tpattern'.` -- EW!
`... hoping it's text matching the pattern '    pattern'.` -- BAD!

 This is doable with `Pattern` objects but  when the pattern is a string, there is no way for `represent_prop` to differentiate r-strings from regular strings.

```python
    @property
    def pattern_to_log(self) -> str:
        """Represent the item in a log-friendly way."""
        return represent_prop(self.pattern)
```
The above will fail the logging tests:
```
Expected :"Text matching the pattern r'(spam)+'."
Actual   :"Text matching the pattern '(spam)+'."
```

To counter this, I considered adding a switch argument to `represent_prop` which would gives the caller the option to turn on/off the `r` in front of strings but I felt like this was an over complicated solution for a single prop.

**Comment by bandophahita** [comment]
@perrygoy back to you.

File: screenpy/resolutions/matches.py (hunk 2 above)
Comment (perrygoy) [reply screenpy/resolutions/matches.py:37]: I still think we should move this logic to `represent_prop`, we have other Pattern-accepting -ables.

Maybe we can do something like this, here:
```python
def pattern_to_log(self):
    """Represent the item in a log-friendly way."""
    loggable_pattern  = represent_prop(self.pattern)
    if not loggable_pattern.startswith("r"):
        loggable_pattern = f"r'{loggable_pattern}'"
    return loggable_pattern
```

How does that sit with you?

Comment (bandophahita) [reply screenpy/resolutions/matches.py:37]: When `self.pattern` is a string, `represent_prop` will call `repr`
```python
    if isinstance(item, str):
        return repr(item)
```
Which is unfortunately not how we want the pattern to appear in the log.

`represent_prop(r"\tpattern")` would return `"\\tpattern"`.  Putting an `r` in front of the string wouldn't fix that.

File: tests/test_resolutions.py [hunk 5]
```diff
@@ -570,12 +572,35 @@ def test_the_test(self) -> None:
         assert m.matches("Spam spam spam spam baked beans and spam")
         assert not m.matches("What do you mean Eugh?!")
 
+    def test_the_test_with_compile(self) -> None:
+        m = Matches(re.compile(r"([Ss]pam ?)+")).resolve()
+
+        assert m.matches("Spam spam spam spam baked beans and spam")
+        assert not m.matches("What do you mean Eugh?!")
```
Comment (perrygoy) [inline tests/test_resolutions.py:579]: I personally wouldn't mind having this be part of the "test the test" test above, that's the pattern we've followed in all the other tests. But i don't mind it being a separate test.

File: tests/test_resolutions.py [hunk 6]
```diff
@@ -570,12 +572,35 @@ def test_the_test(self) -> None:
...
+            r"... hoping it's text matching the pattern r'\tpattern'.",
+            r"    => a string matching '\tpattern'",
+        ]
+
+    def test_description_with_compile(self) -> None:
+        test_match = re.compile(r"(spam)+")
+
+        m = Matches(test_match)
+
+        expected_description = "Text matching the pattern r'(spam)+'."
```
Comment (perrygoy) [inline tests/test_resolutions.py:603]: Same goes for this test; we've just grouped all the "describe correctly" tests in one in other ones (though currently the only examples i can think of are for testing `SeeAllOf` and `SeeAnyOf`).

**Review by perrygoy — APPROVED** [review]
Looks good! I only had a couple comments for consideration, they are not blocking.

**Pull request merged**